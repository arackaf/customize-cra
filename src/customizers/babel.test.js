import {
  addBabelPlugin,
  addBabelPlugins,
  addExternalBabelPlugin,
  addExternalBabelPlugins,
  addDecoratorsLegacy,
  addBabelPreset,
  addBabelPresets,
  fixBabelImports,
  useBabelRc,
  babelInclude,
  babelExclude,
  removeInternalBabelPlugin
} from "./babel";

/**
 * Actually the create-react-app2/3 has two babel-loader rules. One is for
 * app's src and the other is for 'outside of app'( like node_modules ). This
 * config mocks babel-loader for outside of app.
 */
const config = () => ({
  module: {
    rules: [
      {
        oneOf: [
          {
            loader: "babel",
            include: "src",
            options: {
              plugins: []
            }
          },
          {
            loader: "babel",
            exclude: "src",
            options: {}
          }
        ]
      }
    ]
  }
});

test("fixBabelImports adds the babel imports plugin for the provided library", () => {
  const options = { libraryDirectory: "" };
  const plugin = [
    "import",
    { libraryDirectory: "", libraryName: "lodash" },
    "fix-lodash-imports"
  ];
  const actual = fixBabelImports("lodash", options)(config());

  expect(actual).toMatchSnapshot();
});

test("useBabelRc enables the babel loader's babelrc flag", () => {
  const actual = useBabelRc()(config());

  expect(actual).toMatchSnapshot();
});

test("babelInclude sets the babel loader include", () => {
  const include = ["src"];
  const actual = babelInclude(include)(config());

  expect(actual).toMatchSnapshot();
});

test("babelExclude sets the babel loader exclude", () => {
  const exclude = ["fake_directory"];
  const actual = babelExclude(exclude)(config());

  expect(actual).toMatchSnapshot();
});

test("addBabelPlugin returns a function that adds a plugin to the plugins list", () => {
  const plugin = "@babel/plugin-transform-runtime";
  const actual = addBabelPlugin(plugin)(config());

  expect(actual).toMatchSnapshot();
});

test("addExternalBabelPlugin returns a function that adds a plugin to the 'outside of app' babel-loader's plugins list", () => {
  const plugin = "@babel/plugin-proposal-class-properties";
  const actual = addExternalBabelPlugin(plugin)(config());

  expect(actual).toMatchSnapshot();
});

test("addBabelPlugins returns functions that add plugins to the plugins list", () => {
  const plugins = [
    ["@babel/plugin-proposal-object-rest-spread", { loose: true }],
    "@babel/plugin-transform-runtime"
  ];
  const functions = addBabelPlugins(...plugins);
  const actual = functions.reduce((config, fn) => fn(config), config());

  expect(actual).toMatchSnapshot();
});

test("addExternalBabelPlugins returns functions that add plugins to the 'outside of app' babel-loader's plugins list", () => {
  const plugins = [
    ["@babel/plugin-proposal-object-rest-spread", { loose: true }],
    "@babel/plugin-transform-runtime"
  ];
  const functions = addExternalBabelPlugins(...plugins);
  const actual = functions.reduce((config, fn) => fn(config), config());

  expect(actual).toMatchSnapshot();
});

test("addBabelPreset returns a function that adds a preset to the presets list", () => {
  const preset = "@babel/env";
  const conf = config();
  conf.module.rules[0].oneOf[0].options.presets = [];
  const actual = addBabelPreset(preset)(conf);

  expect(actual).toMatchSnapshot();
});

test("addBabelPresets returns functions that add presets to the presets list", () => {
  const presets = [["@babel/env", { loose: true }], "@babel/typescript"];
  const conf = config();
  conf.module.rules[0].oneOf[0].options.presets = [];
  const functions = addBabelPresets(...presets);
  const actual = functions.reduce((config, fn) => fn(config), conf);

  expect(actual).toMatchSnapshot();
});

test("addDecoratorsLegacy returns a function that adds the decorators plugin to the plugins list", () => {
  const actual = addDecoratorsLegacy()(config());

  expect(actual).toMatchSnapshot();
});

test("removeInternalBabelPlugin removes a babel plugin with the matching constructor name", () => {
  function Plugin() {}
  const config = {
    plugins: [new Plugin()]
  };
  const actual = removeInternalBabelPlugin("Plugin")(config);

  expect(actual).toMatchSnapshot();
});
