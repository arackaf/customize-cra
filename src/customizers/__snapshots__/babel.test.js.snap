// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addBabelPlugin returns a function that adds a plugin to the plugins list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [
                "@babel/plugin-transform-runtime",
              ],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`addBabelPlugins returns functions that add plugins to the plugins list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [
                Array [
                  "@babel/plugin-proposal-object-rest-spread",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-transform-runtime",
              ],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`addBabelPreset returns a function that adds a preset to the presets list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [],
              "presets": Array [
                "@babel/env",
              ],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`addBabelPresets returns functions that add presets to the presets list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [],
              "presets": Array [
                Array [
                  "@babel/env",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/typescript",
              ],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`addDecoratorsLegacy returns a function that adds the decorators plugin to the plugins list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [
                Array [
                  "@babel/plugin-proposal-decorators",
                  Object {
                    "legacy": true,
                  },
                ],
              ],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`addExternalBabelPlugin returns a function that adds a plugin to the 'outside of app' babel-loader's plugins list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [
                "@babel/plugin-proposal-class-properties",
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`addExternalBabelPlugins returns functions that add plugins to the 'outside of app' babel-loader's plugins list 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [
                Array [
                  "@babel/plugin-proposal-object-rest-spread",
                  Object {
                    "loose": true,
                  },
                ],
                "@babel/plugin-transform-runtime",
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`babelExclude sets the babel loader exclude 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "exclude": Array [
              "fake_directory",
            ],
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`babelInclude sets the babel loader include 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": Array [
              "src",
            ],
            "loader": "babel",
            "options": Object {
              "plugins": Array [],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`fixBabelImports adds the babel imports plugin for the provided library 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "plugins": Array [
                Array [
                  "import",
                  Object {
                    "libraryDirectory": "",
                    "libraryName": "lodash",
                  },
                  "fix-lodash-imports",
                ],
              ],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;

exports[`removeInternalBabelPlugin removes a babel plugin with the matching constructor name 1`] = `
Object {
  "plugins": Array [],
}
`;

exports[`useBabelRc enables the babel loader's babelrc flag 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "oneOf": Array [
          Object {
            "include": "src",
            "loader": "babel",
            "options": Object {
              "babelrc": true,
              "plugins": Array [],
            },
          },
          Object {
            "exclude": "src",
            "loader": "babel",
            "options": Object {},
          },
        ],
      },
    ],
  },
}
`;
